File: cpp/endpoints/autonomy/goto_poi.cpp
#include <functional>
#include <future>
#include <memory>
#include <string>
#include <yaml-cpp/yaml.h>

#include "clearpath_navigation_msgs/action/execute_go_to_poi.hpp"

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "rclcpp_components/register_node_macro.hpp"

// TODO: Enter your map uuid and your point of interest uuid
const std::string MAP_ID = "";
const std::string POI_ID = "";


namespace goto_poi_action_client
{
class ExecuteGoToPOIActionClient : public rclcpp::Node
{
public:
  using ExecuteGoToPOI = clearpath_navigation_msgs::action::ExecuteGoToPOI;
  using GoalHandleExecuteGoToPOI = rclcpp_action::ClientGoalHandle<ExecuteGoToPOI>;

  explicit ExecuteGoToPOIActionClient(const rclcpp::NodeOptions & options)
  : Node("execute_goto_poi", options)
  {
    YAML::Node robot_config = YAML::LoadFile("/etc/clearpath/robot.yaml");
    namespace_ = robot_config["system"]["ros2"]["namespace"].as<std::string>();

    this->client_ptr_ = rclcpp_action::create_client<ExecuteGoToPOI>(
      this, "/" + namespace_ + "/" + std::string("autonomy/goto_poi"));

    this->timer_ = this->create_wall_timer(
      std::chrono::milliseconds(500),
      std::bind(&ExecuteGoToPOIActionClient::send_goal, this));
  }

  void send_goal()
  {
    using namespace std::placeholders;

    this->timer_->cancel();

    if (!this->client_ptr_->wait_for_action_server()) {
      RCLCPP_ERROR(this->get_logger(), "%s/autonomy/goto_poi action server not available after waiting", namespace_.c_str());
      rclcpp::shutdown();
    }

    auto goal_msg = ExecuteGoToPOI::Goal();
    goal_msg.map_uuid = MAP_ID;
    goal_msg.poi_uuid = POI_ID;

    auto send_goal_options = rclcpp_action::Client<ExecuteGoToPOI>::SendGoalOptions();

    send_goal_options.goal_response_callback = [this](const GoalHandleExecuteGoToPOI::SharedPtr & goal_handle)
    {
      if (!goal_handle) {
        RCLCPP_ERROR(this->get_logger(), "GoTo POI Goal was rejected by server");
      } else {
        RCLCPP_INFO(this->get_logger(), "GoTo POI Goal accepted by server, waiting for result");
      }
    };

    send_goal_options.feedback_callback = [this](
      GoalHandleExecuteGoToPOI::SharedPtr,
      const std::shared_ptr<const ExecuteGoToPOI::Feedback> feedback)
    {
      RCLCPP_INFO_THROTTLE(this->get_logger(), *this->get_clock(), 60000, "GoTo has been running for: %.3f", feedback->elapsed_time);
    };

    send_goal_options.result_callback = [this](const GoalHandleExecuteGoToPOI::WrappedResult & result)
    {
      switch (result.code) {
        case rclcpp_action::ResultCode::SUCCEEDED:
          if (result.result->success)
          {
            RCLCPP_INFO(this->get_logger(), "[GoTo POI Goal Result] Succeeded! Result: Robot arrived at POI successfully!");
          }
          else
          {
            RCLCPP_ERROR(this->get_logger(), "[GoTo POI Goal Result] Succeeded! Result: Robot failed to arrive to POI");
          }
          break;
        case rclcpp_action::ResultCode::ABORTED:
          RCLCPP_ERROR(this->get_logger(), "[GoTo POI Goal Result] Aborted!");
          return;
        case rclcpp_action::ResultCode::CANCELED:
          RCLCPP_ERROR(this->get_logger(), "[GoTo POI Goal Result] Cancelled!");
          return;
        default:
          RCLCPP_ERROR(this->get_logger(), "[GoTo POI Goal Result] Finished with unknown code");
          return;
      }
      rclcpp::shutdown();
    };

    this->client_ptr_->async_send_goal(goal_msg, send_goal_options);
  }

private:
  rclcpp_action::Client<ExecuteGoToPOI>::SharedPtr client_ptr_;
  rclcpp::TimerBase::SharedPtr timer_;
  std::string namespace_;

};  // class ExecuteGoToPOIActionClient

}  // namespace goto_poi_action_client

int main()
{
}

RCLCPP_COMPONENTS_REGISTER_NODE(goto_poi_action_client::ExecuteGoToPOIActionClient)

File: python/endpoints/autonomy/goto_poi.py
from action_msgs.msg import GoalStatus  # Import for reference to status constants
from clearpath_config.common.utils.yaml import read_yaml
from clearpath_navigation_msgs.action import ExecuteGoToPOI
import rclpy
from rclpy.action import ActionClient
from rclpy.node import Node

ROBOT_CONFIG_PATH = '/etc/clearpath/robot.yaml'
LOGGING_NAME = 'ExecuteGoToPOI'

# TODO: Enter your map uuid and point of interest uuid
MAP_ID: str = ''
POI_ID: str = ''


class ExecuteGoToPOIActionClient(Node):

    def __init__(self, namespace: str):
        super().__init__('goto_poi_action_client')
        self._namespace = namespace
        self._action_client = ActionClient(self, ExecuteGoToPOI, f'/{self._namespace}/autonomy/goto_poi')
        self._goto_goal_handle = None
        self._goto_goal_cancelled = False
        self._goto_running = False
        self._goto_completed = False

    #############################################
    # GoTo POI Action Client Functions #
    #############################################
    def send_goto_poi_goal(self, map_id, poi_id):
        # populate goal message
        goal_msg = ExecuteGoToPOI.Goal()
        goal_msg.map_uuid = map_id
        goal_msg.poi_uuid = poi_id

        if not self._action_client.wait_for_server(timeout_sec=5.0):
            self.get_logger().error(f'[{LOGGING_NAME}] /{self._namespace}/autonomy/goto_poi action server not available!')
            self.destroy_node()
            rclpy.shutdown()
            return

        self._send_goal_future = self._action_client.send_goal_async(goal_msg, feedback_callback=self.goto_poi_feedback_cb)
        self._send_goal_future.add_done_callback(self.goto_poi_goal_response_cb)

    def goto_poi_goal_response_cb(self, future):
        self._goto_poi_goal_handle = future.result()
        if not self._goto_poi_goal_handle.accepted:
            self.get_logger().info(f'[{LOGGING_NAME}] GoTo POI Goal rejected')
            self._goto_poi_goal_handle = None
            self.destroy_node()
            rclpy.shutdown()
            return

        self._goto_running = True
        self._goto_completed = False
        self._goto_goal_cancelled = False
        self.get_logger().info(f'[{LOGGING_NAME}] GoTo POI Goal accepted')

        self._get_result_future = self._goto_poi_goal_handle.get_result_async()
        self._get_result_future.add_done_callback(self.goto_poi_get_result_cb)

    def goto_poi_feedback_cb(self, feedback_msg):
        feedback = feedback_msg.feedback
        self.get_logger().info(f'[{LOGGING_NAME}] GoTo POI has been running for: {feedback.elapsed_time:.2f}s', throttle_duration_sec=60)

    def goto_poi_cancel_response_cb(self, future):
        cancel_response = future.result()
        self.get_logger().info(f'[{LOGGING_NAME}] Cancel response received: {cancel_response.return_code}')
        if len(cancel_response.goals_canceling) > 0:
            self.get_logger().info(f'[{LOGGING_NAME}] Canceling of goto poi goal complete')
        else:
            self.get_logger().warning(f'[{LOGGING_NAME}] GoTo POI Goal failed to cancel')

    def goto_poi_get_result_cb(self, future):
        result = future.result().result
        status = future.result().status

        # Compare the status integer against the constants defined in the message
        if status == GoalStatus.STATUS_SUCCEEDED:
            if result.success:
                self.get_logger().info(f'[{LOGGING_NAME}] [GoTo POI Goal Result] Succeeded! Result: Robot completed goto!')
                self._goto_completed = True
            else:
                self.get_logger().info(f'[{LOGGING_NAME}] [GoTo POI Goal Result] Succeeded! Result: Robot failed to complete goto')

        elif status == GoalStatus.STATUS_CANCELED:
            self.get_logger().info(f'[{LOGGING_NAME}] [GoTo POI Goal Result] Cancelled!')
            self._goto_goal_cancelled = True

        elif status == GoalStatus.STATUS_ABORTED:
            self.get_logger().info(f'[{LOGGING_NAME}] [GoTo POI Goal Result] Aborted!')
        else:
            self.get_logger().info(f'[{LOGGING_NAME}] [GoTo POI Goal Result] Finished with unknown status: {status}')

        self._goto_goal_handle = None
        self._goto_running = False

        self.destroy_node()
        rclpy.shutdown()


def main(args=None):
    rclpy.init(args=args)
    robot_config = read_yaml(ROBOT_CONFIG_PATH)
    namespace = robot_config['system']['ros2']['namespace']

    goto = ExecuteGoToPOIActionClient(namespace)
    goto.send_goto_poi_goal(MAP_ID, POI_ID)
    rclpy.spin(goto)


if __name__ == '__main__':
    main()
